rfcs/0000-ckb-light-client/0000-ckb-light-client.md

```diff
@@ -0,0 +1,507 @@
+
+```
+# An MMR with 11 leaves:
+          14
+       /       \
+     6          13
+   /   \       /   \
+  2     5     9     12     17
+ / \   /  \  / \   /  \   /  \
+0   1 3   4 7   8 10  11 15  16 18
+
+# After insertion of a new leaf:
+
+          14
+       /       \
+     6          13            21
+   /   \       /   \         /   \
+  2     5     9     12     17     20
+ / \   /  \  / \   /  \   /  \   /  \
+0   1 3   4 7   8 10  11 15  16 18  19
+```
+
+#### Merkle Root
+
+An MMR is constructed by one or more sub merkle trees (or mountains). Each
+sub merkle tree's root is a peak in MMR, we calculate the MMR root by
+bagging these peaks from left to right.
+
+For example:
+- In the above 11 leaf MMR we have 3 peaks: `14, 17, 18`, we bag these peaks
+  from left to right to get the root:
+  `merge(merge(mmr[14], mmr[17]), mmr[18])`.
+- In the above 12 leaf MMR we have 2 peaks: `14, 21`, we bag these peaks
+  from left to right to get the root: `merge(mmr[14], mmr[21])`.
+
+#### Merkle Proof
+
+The merkle proof is an array of hashes constructed with the following parts:
+- A merkle proof from the leaf's sibling to the peak that contains the leaf.
+- A hash that bags all right-hand side peaks, skip this part if no
+  right-hand peaks.
+- Hashes of all left-hand peaks from right to left, skip this part if no
+  left-hand peaks.
+
+We can reconstruct the merkle root from the proofs.
+
+### The FlyClient Sampling Protocol Under Variable Difficulty
+
+The client uses $g(x) = \frac{1}{(1-x)\ln{\delta}}$ as the probability
+density function (PDF) to sample blocks.
+
+In this probability, $x$ denotes the relative aggregate difficulty weight
+and $\delta$ denotes the relative difficulty weight of the blocks which are
+sampled with probability $1$.
+
+Concretely, let $\delta = c ^ {k}$, where $c$ denote the fraction of the
+adversary's computing power relative to the honest computing power, $p$
+denote the probability of catching the adversary with a single sample and
+$k = \frac{1}{p}$.
+
+The sampling domain is restricted from $0$ to $1 − \delta$ and the blocks in
+the last $\delta$ fraction region should always be checked directly.
+
+Let $n$ denote the number of the blocks in the region which requires
+verifying and $L$ denote the number of blocks in the last $\delta$ fraction
+region. Obviously, we have $L = \delta \times n = c ^ {k} \times n$.
+Accordingly, $k = \log_{c}{(\frac{L}{n})}$.
+
+Define $p_{m} = (1 - \frac{1}{k})^{m}$ as the probability of failure, i.e.,
+not catching the optimal adversary after check $m$ blocks. If we want
+$p_{m} \le 2^{-\lambda}$, then
+$m \ge \frac{\lambda}{\log_{\frac{1}{2}} {(1 - \frac{1}{k})}}$.
+
+## Specification
+
+### MMR Node Specification
+
+We treat all blocks as the leaf nodes of an MMR, and its block number is the
+index of leaves in that MMR.
+
+#### MMR Node
+
+Each MMR node is defined as follows:
+
+- `hash`
+  - For a leaf node, it's an empty hash (`0x0000...0000`).
```

Comment:
If all the leaves are empty hashes, why bother creating the tree? Given a specified number of leaves, the tree is deterministic.

Reply:
This is correct, but the description for non-leaf nodes is confused, I will update it.

How about the description as following?

```
A hash for a node is the hash of its concatenate serialized data.
- For a leaf node, its "hash" field is an empty hash (`0x0000...0000`), since it has no children.
- For a non-leaf node, its "hash" field the hash of its child nodes' hashes (concatenate serialized data of two hashes).
```

Reply:
I think the confusion comes from the bad field name `hash`. A better name could be `children_hash`.

Reply:
Updated!